src/midi/midi_fluidsynth.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2002-2011  The DOSBox Team
+ *  Copyright (C) 2020       Nikos Chantziaras <realnc@gmail.com>
+ *  Copyright (C) 2020       The dosbox-staging team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "midi_fluidsynth.h"
+
+#if C_FLUIDSYNTH
+
+#include <clocale>
+#include <string>
+
+#include "control.h"
+
+// Fluidsynth uses libinstpatch which as of v1.1.3 has the nasty behavior of
+// changing the locale under our nose. We remember and restore the current
+// locale whenever this could happen.
+//
+// https://github.com/realnc/dosbox-core/issues/7
+// https://github.com/swami/libinstpatch/issues/37
+//
+class LocaleGuard final {
+public:
+	~LocaleGuard() { setlocale(LC_ALL, locale_.c_str()); }
+
+private:
+	std::string locale_ = setlocale(LC_ALL, nullptr);
+};
+
+MidiHandlerFluidsynth MidiHandlerFluidsynth::instance;
+
+void init_fluid_dosbox_settings(Section_prop &secprop)
+{
+	constexpr auto when_idle = Property::Changeable::WhenIdle;
+
+	auto *str_prop = secprop.Add_string("fluid.soundfont", when_idle, "");
+	str_prop->Set_help("Soundfont to use with FluidSynth."
+	                   "One must be specified.");
+
+	auto *int_prop = secprop.Add_int("fluid.samplerate", when_idle, 44100);
+	int_prop->SetMinMax(8000, 96000);
+	int_prop->Set_help("Sample rate (Hz) Fluidsynth will render at. (min "
+	                   "8000, max 96000");
+
```

Comment:
Suggest we configure Fluid's rendering rate directly from the mixer's rate, given Fluid's stream will be fed into the mixer and resampled back to the mixer's rate if it doesn't match (resulting in unnecessary resampling, and not very good resampling at that versus "best in class" resampling libraries like SSRC or the Speex Resampler).

For example:
 - if Fluid is configured less than the mixer's rate, then the mixer will resample the stream upward; in which case, the user should just set the mixer to a lower rate anyway - and avoid the double-resampling. 
 - If Fluid is configured with a high rate - say 88200 or 96000, the mixer will clobber that fidelity back down to 44100 (if that's what it set at), and add any resampling artifacts while it's at it.

The only way to eliminate mixer's resampling artifacts from affecting Fluid's stream is to make the two rates match.. so instead, we can do this automatically by simply making Fluid render at the mixer's rate.. so we're always passing through "native" samples to the OS'es audio subsystem.

Reply:
Agreed; I kept it in because it was part of the original patch, also - this might be a topic for subsequent PR - you're already getting rid of `gusrate` in GUS PR, but does the same apply to `sbrate` as well?

Reply:
>     * If Fluid is configured with a high rate - say 88200 or 96000, the mixer will clobber that fidelity back down to 44100 (if that's what it set at), and add any resampling artifacts while it's at it.

Note that if you have (or will get) a high quality resampler in dosbox-staging, allowing for running fsynth at a higher sample rate and then downsampling to the output rate can provide some quality improvements, as a higher internal rate reduces soundfont instrument resampling artifacts. Some synths do that by default (it's called "oversampling".)

It's a minor thing in general though and probably most users won't notice.

Reply:
Resampling is always harmful when everything else is held equal:
 - Resampling always adds CPU time (because it needs to pass through the original samples)
 - Resampling always adds latency; the lower the latency the larger the cost in artifacts added CPU consumption

When the source material **_can maximize its information content as a function of frequency_**, then it always makes sense to generate directly to the end-state output frequency instead of oversampling at the source and then downsampling to the end-state output frequency.

The caveat is when the source material **_can only maximize its information as a function of one or more fixed frequencies_**.   Given we know resampling is harmful, the question becomes: is running the source at it's non-prefered frequency more harmful than resampling it's prefered frequency.


> you're already getting rid of gusrate in GUS PR, but does the same apply to sbrate as well

For adlib, if all it's doing it sampling it's source "FM waveform" at the desired output sampling rate, then yes - we would want the adlib to generate at the mixer's output to avoid dosbox's resampler (instead of oversampling in the adlib and then damaging the signal further resampling back down to the mixer's output rate).

> allowing for running fsynth at a higher sample rate and then downsampling to the output rate can provide some quality improvements

If oversampling to say 96000 is only just mathematically interpolating at a higher output rate, then resampling that back down to 44100 or 48000 (no matter how good the resampler is), will still be worse than simply having fluidsynth generate at the end-rate directly, and avoid the resampling.

However, if fuildsynth's math calculations assume a 96000 output rate (and break down in various way at other rates) -- then yes, it would make sense to run fluidsyth at 96000 and then resample down to to the playback rate.

Reply:
Wanted follow up on the Adlib question: 
1. Should we run the Adlib OPL emulator at 49716 Hz and then resample back down to the mixer's output rate or, 2. Should we run the Adlib OPL emulator at the mixer's output rate (eliminating the CPU cost of the higher OPL rate as well as the CPU cost and latency of resampling).

I took two recordings

1. `oplrate = 49716` Hz resampled back down to 16000 Hz.  
2. `oplrate = 16000` Hz, saved natively at 16000 Hz. 

I was unable to see any differences in the wave forms (zoomed in or out in Audacity), here's the first second of audio:

![2020-08-15_12-57](https://user-images.githubusercontent.com/1557255/90320524-fc092400-def6-11ea-92e8-8e723161443c.png)

![2020-08-15_12-58](https://user-images.githubusercontent.com/1557255/90320525-fca1ba80-def6-11ea-9c65-d5977f3b69b9.png)

Likewise I wasn't able to hear any differences when listening to them with decent Sennheiser headphones, and repeating sections over and over.

I'm a human though.. perhaps the first recording still contains more information that my eyes and ears aren't differentiating.

I used FLAC compression to measure their information content (or entropy), which from [information theory](https://en.wikipedia.org/wiki/Entropy_(information_theory)#Data_compression) says that the entropy rate of a data source can be measured by the average number of bits per symbol needed to encode it.

FLAC was able to compress both recordings to the same reported level of compression and inspecting the actual byte-quantities in the resulting FLAC files differed by a mere 0.039%! Wow. 

So this answers it.. we're not losing any information by running OPL at the mixer direct output rate, and in doing so we avoid:
- wasting CPU cycles by over-sampling the OPL
- wasting CPU cycles by resampling the FM channel back down in the mixer
- introducing resampling artifacts in the output stream

Based on this, I suggest we run the OPL at `mixer.rate`.